source/common/upstream/health_checker_impl.cc

```diff
@@ -145,6 +146,35 @@ void HealthCheckerImplBase::runCallbacks(HostSharedPtr host, bool changed_state)
   }
 }
 
+void HealthCheckerImplBase::HealthCheckerSinkImpl::setUnhealthy() {
+  // This is called cross thread. The cluster/health checker might already be gone.
+  std::shared_ptr<HealthCheckerImplBase> health_checker = health_checker_.lock();
+  if (health_checker) {
+    health_checker->setUnhealthyCrossThread(host_.lock());
+  }
+}
+
+void HealthCheckerImplBase::setUnhealthyCrossThread(const HostSharedPtr& host) {
+  // The threading here is complex. The cluster owns the only strong reference to the health
+  // checker. It might go away when we post to the main thread. We capture a weak reference and
+  // make sure it is still valid when we get to the other thread. Additionally, the host/session
+  // may also be gone by then so we check that also.
+  std::weak_ptr<HealthCheckerImplBase> weak_this = shared_from_this();
+  dispatcher_.post([weak_this, host]() -> void {
+    std::shared_ptr<HealthCheckerImplBase> shared_this = weak_this.lock();
```

Comment:
This logic seems complicated. Some of this is inherent due to ownership structure, but I wonder if it could be simplified by having the main thread (in the post body) do the cluster -> health checker resolution, and only passing in the cluster? Or host+cluster -> health checker resolution?

The weak pointers seem valid, but it becomes fairly hard to reason about the combination of shared_ptr and then multiple weak_ptrs in the sink.

Reply:
Again FWIW this is basically identical to logic we do for outlier detection: https://github.com/lyft/envoy/blob/master/source/common/upstream/outlier_detection_impl.cc#L220

I agree the logic is complicated, but I'm not quite sure how to make it simpler. Even if we pass a cluster, we need to pass a weak_ptr, because the code relies on clusters going away on the main thread immediately. I figured it was better to have all the shared_ptr/weak_ptr logic internal to this thing like we did in outlier detector? Let me try adding some more comments.